experiments/repo_request.py

```diff
@@ -0,0 +1,108 @@
+import pandas as pd
+import numpy as np
+import os
+import requests
+import json
+
+def load_data(filepath):
+    try:
+        df = pd.read_csv(filepath)
+        print(f'Dataframe shape: {df.shape}')
+        print(f'Dataframe columns: {df.columns.tolist()}')
+        df.info()
+        return df
+
+    except Exception as e:
+        print(f'Error Loading data: {e}')
+    return None
+
+def check_and_clean_data(df):
+    # Checking for null values
+    null_counts = df.isnull().sum()
+    if np.any(null_counts):
+        print("Null values found:")
+        print(null_counts[null_counts > 0])
+    else:
+        print("No null values found.")
+
+    # Checking for duplicates
+    duplicates = df.duplicated().sum()
+    if duplicates:
+        print(f"Number of duplicate rows: {duplicates}")
+    else:
+        print("No duplicate rows found.")
+
+def construct_search_request(repo_path):
+    # constructing the search query
+    search_query = f'test in:path -filename:.txt -filename:.md repo:{repo_path}'
+    payload = {'q': search_query, 'type': 'code'}
+    # The URL for the Search Code endpoint
+    search_url = f'https://api.github.com/search/code'
+    return payload, search_url
+
+def extract_reponame_and_username(repo_path):
+    parts = repo_path.split('/')
+    print(parts)
+    repo_path = '/'.join(parts[-2:])  # This joins the last two parts using '/'
+    return repo_path
+
+def main():
+    data_path = 'df'
+    # Loading the dataframe
+    df = load_data(data_path)
+
+    # Checking the Null and duplicate values
+    check_and_clean_data(df)
+
+    # I will keep the first occurrence of each duplicate row and remove the others:
+    df = df.drop_duplicates(keep='first')
+
+    # I will only consider github.com domain
+    github_df = df[df['repourl'].str.contains('github.com')]
+    df.to_csv('github_df', index = False)
+
+    # Accessing the environmental variable (PAT)
+    pat = os.getenv('MY_PAT')
+
+    # Make the authenticated request
+    headers = {
+        'Authorization' : f'token {pat}',
+        'X-GitHub-Api-Version': '2022-11-28'
+               }
+    # Some of the URLs end with "/". I need to remove them.
```

Comment:
Why do they need removing? be good to explain briefly why they need removing in the comment.